mava/systems/sac/anakin/ff_hasac.py

```diff
@@ -0,0 +1,717 @@
+        next_q2_val = q_net.apply(params.q.targets.q2, data.next_obs, joint_next_actions)
+        next_q_val = jnp.minimum(next_q1_val, next_q2_val)
+        next_q_val = next_q_val - jnp.exp(params.log_alpha) * next_log_prob
+
+        target_q_val = data.reward + (1.0 - data.done) * cfg.system.gamma * next_q_val  # (B, A, 1)
+
+        # Update Q function.
+        q_grad_fn = jax.grad(q_loss_fn, has_aux=True)
+        q_grads, q_loss_info = q_grad_fn(params.q.online, data.obs, data.action, target_q_val)
+        # Mean over the device and batch dimension.
+        q_grads, q_loss_info = lax.pmean((q_grads, q_loss_info), axis_name="device")
+        q_grads, q_loss_info = lax.pmean((q_grads, q_loss_info), axis_name="batch")
+        q_updates, new_q_opt_state = q_opt.update(q_grads, opt_states.q)
+        new_online_q_params = optax.apply_updates(params.q.online, q_updates)
+
+        # Target network polyak update.
+        new_target_q_params = optax.incremental_update(
+            new_online_q_params, params.q.targets, cfg.system.tau
+        )
+
+        # Repack params and opt_states.
+        q_and_target = QValsAndTarget(new_online_q_params, new_target_q_params)
+        params = params._replace(q=q_and_target)
+        opt_states = opt_states._replace(q=new_q_opt_state)
+
+        return params, opt_states, q_loss_info
+
+    def update_actor_and_alpha(
+        params: SacParams, opt_states: OptStates, data: Transition, key: chex.PRNGKey
+    ) -> Tuple[SacParams, OptStates, Metrics]:
+        """Update the actor and alpha parameters. Compensated for the delay in policy updates."""
+        alpha_grad_fn = jax.value_and_grad(alpha_loss_fn)
+        actor_grad_fn = jax.value_and_grad(actor_loss_fn)
+
+        # compensate for the delay by doing `policy_frequency` updates instead of 1.
+        assert cfg.system.policy_update_delay > 0, "Need to have a policy update delay > 0."
+        for _ in range(cfg.system.policy_update_delay):
+            key, act_key, agent_order_key = jax.random.split(key, 3)
+            act_keys = jax.random.split(act_key, env.num_agents)
+            if cfg.system.shuffle_agents:
+                agent_ids = jax.random.permutation(agent_order_key, env.num_agents)
+            else:
+                agent_ids = jnp.arange(env.num_agents)
+
+            joint_actions, log_probs = get_actions(
+                params.actor, actor_net, act_keys, env.num_agents, env.action_dim, data.obs
+            )
+
+            # HASAC sequential update: run the normal actor update one at a time instead of batched.
+            # Update the joint actions after updating the actor and use the new joint actions.
+            for agent_id in agent_ids:
+                key, actor_key = jax.random.split(key)
+
+                agent_params = tree_slice(params.actor, agent_id)
+                agent_opt_state = tree_slice(opt_states.actor, agent_id)
+                agent_obs = tree_slice(data.obs, jnp.s_[:, agent_id])
```

Comment:
Similarly, a shape here would be nice. Also what is `jnp.s_`? I haven't seen it before.